contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,345 @@
+1. API Server
+1. API Machinery
+1. Controllers (that wish to utilize this feature)
+1. Kubelet
+1. Kubectl
+
+## Use Cases
+
+1. As the implementor of a controller or operator that manages a stateful 
+workload, I want to send a signal to the Pods created by my controller or 
+operator to control the application's replication behavior during Pod 
+termination.
+1. As the implementor of tenant applications on a Kubernetes cluster, I want 
+my applications to receive a signal when they are evicted from a node due to 
+intolerable taints.
+1. As the administrator of a Kubernetes cluster, I want to provide a signal to 
+tenant applications that they are being terminated in violation of their 
+declared disruption budgets or termination grace periods due to a superseding, 
+global, organization policy (e.g. My organization requires that an application's 
+termination grace period and disruption budget allow managed upgrades to make 
+progress.)
+1. As a user I want to be able to send a termination reason to my Pods when I 
+use the Kubernetes API or kubectl to gracefully delete a Pod.
+1. As a Reliability Engineer for an application running on a Kubernetes cluster,
+I want to configure Pods to communicate the reason that they were terminated via 
+a HTTP GET request so that I can configure monitoring and alerting based on 
+historic trends.
+
+## Assumptions
+
+1. Kubelet will always invoke the pre-stop handler prior to sending a `TERM` 
+signal to the entry point of a container.
+1. pre-stop handlers will not contain complicated or long running business logic. 
+The logic associated with container cleanup will be executed by the container's
+entry point during the container's configured termination grace period.
+
+## Requirements
+
+1. Termination reason must be available to a container at the time its pre-stop 
+handler is invoked. In this way, applications can perform any actions
+necessary to configure the container's entry point prior to the entry point 
+receiving the `TERM` signal.
+1. The delivery mechanism for termination reason must be consumable by, and 
+ compatible with, existing HTTP and command based pre-stop handlers.
+1. Any headers, variables, or parameters used to communicate a termination 
+reason must not be hard coded, and the receiving container must be able to 
+override the defaults.
+
+## API Objects
+
+Termination reason is implemented by a string type.
+
+```golang
+// TerminationReason is a string type used to indicate the reason for 
+// termination to the preStop lifecycle hook of a Pod's containers.
+type TerminationReason string
+
+const (
+	// ReasonEviction is the default reason used to communicate that a Pod has 
+ 	// been terminated due to an eviction.
+	ReasonEviction TerminationReason = "Eviction"
+	// ReasonIntolerableTaint is the default reason used to communicate that a Pod 
+	// has been terminated due to a taint for which it has no declared toleration.
+	ReasonIntolerableTaint = "IntolerableTaint"
```

Comment:
`TerminationIntention` enum seems to be promising since it eliminates the needs for applications to check various reasons and then decide whether/where the pods would be restarted; otherwise, applications will probably need to be updated to handle each newly added reasons in the future.

This value will not be in DeleteOptions since it's only specific to pod deletion and `kubectl` will take this value as an enum.